src/components/coin.ts

```diff
@@ -72,3 +91,76 @@ export const adjustCoinBalanceByUserId = async (userId: string, amount: number):
     amount
   );
 };
+
+/*
+  If (user, bonusType) doesn't exist, create row with current time as this bonusType log.
+  Otherwise, update last_granted to CURRENT_TIMESTAMP.
+*/
+export const updateUserBonusTableByUserId = async (userId: string, bonusType: BonusType): Promise<void> => {
+  const db = await openDB();
+  await db.run(
+    `
+    INSERT INTO user_coin_bonus (user_id, bonus_type, last_granted) VALUES (?, ?, CURRENT_TIMESTAMP)
+    ON CONFLICT(user_id, bonus_type)
+    DO UPDATE SET last_granted = CURRENT_TIMESTAMP`,
+    userId,
+    bonusType
+  );
+};
+
+/*
+  Get the time of the latest bonus applied to a user based on type
+*/
+export const latestBonusByUserId = async (userId: string, type: BonusType): Promise<UserCoinBonus | undefined> => {
+  const db = await openDB();
+  const res: UserCoinBonus | undefined = await db.get(
+    `SELECT last_granted FROM user_coin_bonus WHERE user_id = ? AND bonus_type = ?`,
+    userId,
+    type
+  );
+  return res;
+};
+
+/*
+  Time bonus
+*/
+export const timeBonusByUserId = async (userId: string, bonusType: BonusType): Promise<boolean> => {
+  const bonusOfInterest = coinBonusMap.get(bonusType);
+
+  if (bonusOfInterest === undefined) {
+    return true;
+  }
+
+  const lastBonusOccurence = await latestBonusByUserId(userId, bonusOfInterest.type);
+  if (lastBonusOccurence === undefined) {
+    await adjustCoinBalanceByUserId(userId, bonusOfInterest.amount);
+    await updateUserBonusTableByUserId(userId, bonusType);
+    return true;
+  }
+
+  const lastBonusOccurenceTime = new Date(lastBonusOccurence['last_granted']).getTime();
+  const nowTime = new Date().getTime();
+  const cooldown = nowTime - bonusOfInterest.cooldown;
+
+  if (lastBonusOccurenceTime < cooldown) {
```

Comment:
this block is has the exact same logic as the if statement before, can we just combine these two if statements into 1?